File: Explore Card Track/3. Linked List/0707-design-linked-list/0707-design-linked-list-DoublyLinkedList.py
class Node:
    def __init__(self, val):
        self.prev = None  # Pointer to the previous node
        self.val = val    # Value stored in the node
        self.next = None  # Pointer to the next node

class MyLinkedList:
    def __init__(self):
        self.head = None  # Head pointer to the first node of the list
        self.length = 0   # Length of the linked list

    # Get the value of the node at the given index
    def get(self, index: int) -> int:
        if index >= self.length or index < 0:
            return -1  # Return -1 if the index is invalid
        currentNode = self.head
        while index:
            currentNode = currentNode.next
            index -= 1
        return currentNode.val

    # Add a new node with value `val` at the head of the list
    def addAtHead(self, val: int) -> None:
        newNode = Node(val)
        if self.head is None:  # If the list is empty, just set head to newNode
            self.head = newNode
        else:
            newNode.next = self.head
            self.head.prev = newNode  # Update the previous head's `prev` pointer
            self.head = newNode       # Move the head to the new node
        self.length += 1

    # Add a new node with value `val` at the tail of the list
    def addAtTail(self, val: int) -> None:
        newNode = Node(val)
        if self.head is None:  # If the list is empty, set the head to newNode
            self.head = newNode
        else:
            currentNode = self.head
            while currentNode.next is not None:  # Traverse to the last node
                currentNode = currentNode.next
            currentNode.next = newNode  # Link the last node to the new node
            newNode.prev = currentNode  # Set the `prev` pointer of the new node
        self.length += 1    

    # Add a new node with value `val` at the given index
    def addAtIndex(self, index: int, val: int) -> None:
        if index < 0 or index > self.length:
            return  # Invalid index
        elif index == 0:
            self.addAtHead(val)  # Insert at the head
        elif index == self.length:
            self.addAtTail(val)  # Insert at the tail
        else:
            newNode = Node(val)
            currentNode = self.head
            for _ in range(index - 1):
                currentNode = currentNode.next  # Traverse to the node just before the index
            newNode.next = currentNode.next
            newNode.prev = currentNode
            currentNode.next.prev = newNode  # Update the next node's `prev` pointer
            currentNode.next = newNode       # Link the current node to the new node
            self.length += 1

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.length:
            return  # Invalid index
        elif index == 0:
            # Deleting the head node
            if self.length == 1:
                self.head = None  # If there's only one element, set head to None
            else:
                self.head = self.head.next  # Move head to the next node
                self.head.prev = None       # Set the new head's `prev` to None
        else:
            currentNode = self.head
            for _ in range(index - 1):
                currentNode = currentNode.next  # Traverse to the node just before the index
            nodeToDelete = currentNode.next
            currentNode.next = nodeToDelete.next  # Link the current node to the next of next
            if nodeToDelete.next is not None:
                nodeToDelete.next.prev = currentNode  # Update the `prev` pointer if not deleting the tail
        self.length -= 1

# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)

File: 0707-design-linked-list/0707-design-linked-list-DoublyLinkedList.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    struct Node* prev;  // Pointer to the previous node
    int val;            // Value stored in the node
    struct Node* next;  // Pointer to the next node
} Node;

typedef struct {
    Node* head;    // Pointer to the head node
    Node* tail;    // Pointer to the tail node for efficiency in tail operations
    int length;    // Length of the linked list
} MyLinkedList;

// Function to create a new linked list
MyLinkedList* myLinkedListCreate() {
    MyLinkedList* list = (MyLinkedList*)malloc(sizeof(MyLinkedList));
    if (list == NULL) {
        fprintf(stderr, "Linked List Memory Allocation Failed");
        return NULL;
    }
    list->head = NULL;  // Initialize head and tail to NULL
    list->tail = NULL;
    list->length = 0;
    return list;
}

// Function to get the value at the given index
int myLinkedListGet(MyLinkedList* obj, int index) {
    if (index < 0 || index >= obj->length) {
        return -1;  // Return -1 if the index is invalid
    }
    Node* currentNode = obj->head;
    for (int i = 0; i < index; i++) {
        currentNode = currentNode->next;
    }
    return currentNode->val;
}

// Function to add a node with value `val` at the head
void myLinkedListAddAtHead(MyLinkedList* obj, int val) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        fprintf(stderr, "Node Memory Allocation Failed");
        return;
    }
    newNode->val = val;
    newNode->next = obj->head;  // New node points to the current head
    newNode->prev = NULL;       // Since it's the head, prev is NULL

    if (obj->head != NULL) {
        obj->head->prev = newNode;  // Set the current head's prev to the new node
    }
    
    obj->head = newNode;  // Update the head to be the new node
    if (obj->tail == NULL) {
        obj->tail = newNode;  // If the list was empty, the new node is also the tail
    }
    obj->length++;
}

// Function to add a node with value `val` at the tail
void myLinkedListAddAtTail(MyLinkedList* obj, int val) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        fprintf(stderr, "Node Memory Allocation Failed");
        return;
    }
    newNode->val = val;
    newNode->next = NULL;  // Since it's the tail, next is NULL
    newNode->prev = obj->tail;  // New node's prev points to the current tail

    if (obj->tail != NULL) {
        obj->tail->next = newNode;  // Update current tail's next to the new node
    }
    
    obj->tail = newNode;  // Update the tail to be the new node
    if (obj->head == NULL) {
        obj->head = newNode;  // If the list was empty, the new node is also the head
    }
    obj->length++;
}

// Function to add a node with value `val` at the given index
void myLinkedListAddAtIndex(MyLinkedList* obj, int index, int val) {
    if (index < 0 || index > obj->length) {
        return;  // Invalid index
    }
    if (index == 0) {
        myLinkedListAddAtHead(obj, val);  // Insert at the head
    } else if (index == obj->length) {
        myLinkedListAddAtTail(obj, val);  // Insert at the tail
    } else {
        Node* newNode = (Node*)malloc(sizeof(Node));
        if (newNode == NULL) {
            fprintf(stderr, "Node Memory Allocation Failed");
            return;
        }
        newNode->val = val;
        
        Node* currentNode = obj->head;
        for (int i = 0; i < index - 1; i++) {
            currentNode = currentNode->next;  // Traverse to the node before the index
        }
        
        newNode->next = currentNode->next;  // New node points to the node currently at index
        newNode->prev = currentNode;        // New node's prev points to the current node
        
        if (currentNode->next != NULL) {
            currentNode->next->prev = newNode;  // Update the next node's prev pointer
        }
        currentNode->next = newNode;  // Link the current node to the new node
        
        obj->length++;
    }
}

// Function to delete the node at the given index
void myLinkedListDeleteAtIndex(MyLinkedList* obj, int index) {
    if (index < 0 || index >= obj->length) {
        return;  // Invalid index
    }
    
    Node* currentNode = obj->head;
    if (index == 0) {
        // Deleting the head
        obj->head = currentNode->next;
        if (obj->head != NULL) {
            obj->head->prev = NULL;  // Set the new head's prev to NULL
        } else {
            obj->tail = NULL;  // List becomes empty, so tail should be NULL
        }
        free(currentNode);
    } else {
        for (int i = 0; i < index; i++) {
            currentNode = currentNode->next;  // Traverse to the node at the index
        }
        
        if (currentNode->prev != NULL) {
            currentNode->prev->next = currentNode->next;  // Update previous node's next pointer
        }
        
        if (currentNode->next != NULL) {
            currentNode->next->prev = currentNode->prev;  // Update next node's prev pointer
        } else {
            obj->tail = currentNode->prev;  // Update tail if we are deleting the last node
        }
        
        free(currentNode);
    }
    
    obj->length--;
}

// Function to free the linked list and its nodes
void myLinkedListFree(MyLinkedList* obj) {
    Node* currentNode = obj->head;
    while (currentNode != NULL) {
        Node* nextNode = currentNode->next;
        free(currentNode);  // Free each node
        currentNode = nextNode;
    }
    free(obj);  // Free the linked list structure
}

/**
 * Your MyLinkedList struct will be instantiated and called as such:
 * MyLinkedList* obj = myLinkedListCreate();
 * int param_1 = myLinkedListGet(obj, index);
 * myLinkedListAddAtHead(obj, val);
 * myLinkedListAddAtTail(obj, val);
 * myLinkedListAddAtIndex(obj, index, val);
 * myLinkedListDeleteAtIndex(obj, index);
 * myLinkedListFree(obj);
 */
